packages/docprovider/src/ydrive.ts

```diff
@@ -0,0 +1,234 @@
+
+  /**
+   * Save a file.
+   *
+   * @param localPath - The desired file path.
+   *
+   * @param options - Optional overrides to the model.
+   *
+   * @returns A promise which resolves with the file content model when the
+   *   file is saved.
+   *
+   * #### Notes
+   * Ensure that `model.content` is populated for the file.
+   *
+   * Uses the [Jupyter Notebook API](http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/notebook/master/notebook/services/api/api.yaml#!/contents) and validates the response model.
+   */
+  async save(
+    localPath: string,
+    options: Partial<Contents.IModel> = {}
+  ): Promise<Contents.IModel> {
+    if (this._sharedPaths.has(localPath)) {
+      // Save is done from the backend
+      return this.get(localPath, { ...options, content: false });
+    } else {
+      return super.save(localPath, options);
+    }
+  }
+
+  private _onCreate = (
+    options: IFactory.IOptions,
+    sharedModel: YDocument<DocumentChange>
+  ) => {
+    try {
+      const provider = new WebSocketProvider({
+        url: URLExt.join(this.serverSettings.wsUrl, Y_DOCUMENT_PROVIDER_URL),
+        path: options.path,
+        format: options.format,
+        contentType: options.contentType,
+        model: sharedModel,
+        user: this._user
+      });
+
+      const key = `${options.contentType}:${options.format}:${options.path}`;
+      this._providers.set(key, provider);
+      // FIXME
+      sharedModel.disposed.connect(() => {
+        const provider = this._providers.get(key);
+        if (provider) {
+          provider.dispose();
+          this._providers.delete(key);
+        }
+      });
+    } catch (error) {
+      // Falling back to the contents API if opening the websocket failed
+      //  This may happen if the shared document is not a YDocument.
+      console.error(
+        `Failed to open websocket connection for ${options.path}.\n:${error}`
+      );
+    }
+  };
+
+  private _user: User.IManager;
+  private _sharedPaths: Set<string>;
+  private _providers: Map<string, WebSocketProvider>;
+}
+
+/**
+ *
+ */
+export class SharedModelFactory implements IFactory {
+  constructor(
+    private _onCreate: (
+      options: IFactory.IOptions,
+      sharedModel: YDocument<DocumentChange>
+    ) => void
+  ) {}
+
+  createNew(options: IFactory.IOptions): ISharedDocument | undefined {
+    if (
+      typeof options?.format !== 'string' ||
+      typeof options?.contentType !== 'string'
+    ) {
+      throw new Error('Format and content type must be provided.');
+    }
+
+    if (options.collaborative !== true) {
+      return;
+    }
+
+    let sharedModel: YDocument<DocumentChange> | undefined;
+    switch (options?.contentType) {
+      case 'file':
+        sharedModel = new YFile();
+        break;
+      case 'notebook':
+        sharedModel = new YNotebook();
+        break;
+      case 'directory':
+      default:
+        return;
+    }
+
+    this._onCreate(options, sharedModel);
```

Comment:
Let's prepare the future.

```suggestion
//      default: // FIXME we should request a registry for the proper sharedModel
    }

    if (sharedModel) {
      this._onCreate(options, sharedModel);
    }
```